### PR TITLE
Make sure we return the task ref when raising task_cancelled

### PR DESCRIPTION
**Pull request by johnelse**
Also use TaskHelpers.exn_if_cancelling wherever possible.

### CODE REVIEW

File: ocaml/xapi/cli_operations.ml [hunk 1]
```diff
@@ -3195,7 +3195,8 @@ let vm_import fd printer rpc session_id params =
 														* alert it to the failure. *)
 
 														(let l=Client.Task.get_current_operations rpc session_id importtask in
-															if List.exists (fun (_,x) -> x=`cancel) l then raise (Api_errors.Server_error(Api_errors.task_cancelled,[])));
+															if List.exists (fun (_,x) -> x=`cancel) l
```
Comment (jonludlam) [inline ocaml/xapi/cli_operations.ml]: Grief, I thought this was just blank lines until I scrolled over to the right. This function could do with some refactoring, methinks.

File: ocaml/xapi/xapi_vm_snapshot.ml [hunk 2]
```diff
@@ -219,7 +219,9 @@ let checkpoint ~__context ~vm ~new_name =
 			Xapi_xenops.resume ~__context ~self:vm ~start_paused:false ~force:false;
 		end;
 		match snap with
-		| None      -> raise (Api_errors.Server_error (Api_errors.task_cancelled,[]))
+		| None  ->
+			let task_id = Context.get_task_id __context in
+			raise Api_errors.(Server_error (task_cancelled,[Ref.string_of task_id]))
```
Comment (jonludlam) [inline ocaml/xapi/xapi_vm_snapshot.ml]: exn_if_cancelling?

File: ocaml/idl/ocaml_backend/taskHelper.ml [hunk 3]
```diff
@@ -176,7 +176,11 @@ let is_cancelling ~__context =
 	  List.exists (fun (_,x) -> x=`cancel) l
 
 let exn_if_cancelling ~__context =
-  if is_cancelling ~__context then raise (Api_errors.Server_error (Api_errors.task_cancelled,[]))
+	if is_cancelling ~__context
+	then begin
+		let task_id = Context.get_task_id __context in
+		raise Api_errors.(Server_error (task_cancelled, [Ref.string_of task_id]))
```
Comment (jonludlam) [inline ocaml/idl/ocaml_backend/taskHelper.ml]: Perhaps we could factor out these two lines, as they're repeated in several places below

File: ocaml/xapi/xapi_vm_snapshot.ml (hunk 2 above)
Comment (johnelse) [inline ocaml/xapi/xapi_vm_snapshot.ml]: If snap = None then either (a) the checkpoint task was cancelled during the suspend or (b) the VM clone was cancelled. Using exn_if_cancelling at this point would catch (a) but not (b).

File: ocaml/xapi/cli_operations.ml (hunk 1 above)
Comment (johnelse) [inline ocaml/xapi/cli_operations.ml]: Yuh :(

File: ocaml/idl/ocaml_backend/taskHelper.ml (hunk 3 above)
Comment (johnelse) [inline ocaml/idl/ocaml_backend/taskHelper.ml]: exn_if_cancelling would be more reusable if it took a cleanup function - there are a few bits of code of the form

```
if is_cancelling(task):
  do_some_cleanup()
  raise task_cancelled
```

I'll see if I can tidy things up a bit.

**Comment by johnelse** [comment]
PR updated.

File: ocaml/xapi/xapi_templates_install.ml [hunk 4]
```diff
@@ -71,7 +71,8 @@ let post_install_script rpc session_id __context install_vm vm (script, vbds) =
 		   begin
 		     Unix.kill (Forkhelpers.getpid pid) Sys.sigterm;
 		     let _ = Forkhelpers.waitpid pid in
-		     raise (Api_errors.Server_error (Api_errors.task_cancelled, []))
+		     let task_id = Context.get_task_id __context in
+		     raise Api_errors.(Server_error (task_cancelled, [Ref.string_of task_id]))
 		   end;
```
Comment (jonludlam) [inline ocaml/xapi/xapi_templates_install.ml]: Couldn't this bit use 'raise_cancelled'?

Comment (johnelse) [inline ocaml/xapi/xapi_templates_install.ml]: Oops, yes, I'll fix that.

**Comment by jonludlam** [comment]
Looks good now!

**Pull request merged**